include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+			{
+				m_container.insert(std::move(entry));
 			}
 
-			virtual IShader::E_CONTENT_TYPE getCodeContentType() const { return IShader::E_CONTENT_TYPE::ECT_UNKNOWN; };
+			// can move to .cpp and have it not inline
+			inline core::smart_refctd_ptr<asset::ICPUShader> find(const SEntry& mainFile, CIncludeFinder* finder)
+			{
+				auto foundRange = m_container.equal_range(mainFile);
+				for (auto& found = foundRange.first; found != foundRange.second; found++)
+				{
+					bool allDependenciesMatch = true;
+					// go through all dependencies
+					for (auto i = 1; i != found->dependencies.size(); i++)
+					{
+						const auto& dependency = found->dependencies[i];
+
+						IIncludeLoader::found_t header;
+						if (dependency.standardInclude)
+							header = finder->getIncludeStandard(dependency.requestingSourceDir, dependency.identifier);
+						else
+							header = finder->getIncludeRelative(dependency.requestingSourceDir, dependency.identifier);
+
+						if (header.hash != dependency.hash || header.contents != dependency.contents)
+						{
+							allDependenciesMatch = false;
+							break;
+						}
+					}
+					if (allDependenciesMatch) {
+						if (!found->value) { // Load shader if not already loaded
+							// If the Cache has no storageBuffer, then it's fresh new (not picked up from serialization): all compiled shaders should be loaded
+							// If codeByteSize is 0, then the Entry found lives in the Cache but not in the serialization, so the Entry was generated in the current runtime:
+							// This means the shader should be loaded
+							assert(storageBuffer && found->shaderParams.codeByteSize != 0);
+							auto code = core::make_smart_refctd_ptr<ICPUBuffer>(found->shaderParams.codeByteSize);
+							memcpy(code->getPointer(), (uint8_t*)storageBuffer->getPointer() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize + found->shaderParams.offset, found->shaderParams.codeByteSize);
+							found->value = core::make_smart_refctd_ptr<ICPUShader>(std::move(code), found->shaderParams.stage, found->shaderParams.contentType, std::move(std::string(found->shaderParams.filepathHint)));
+						}
+						return found->value;
+					}	
+				}
+				return nullptr;
+			}
+
+			CCache() {}
+
+			// TODO: add methods as needed, e.g. to serialize and deserialize to/from a pointer
+			std::vector<uint8_t> serialize();
+			static core::smart_refctd_ptr<CCache> deserialize(core::smart_refctd_ptr<ICPUBuffer> serializedCache);
+			static core::smart_refctd_ptr<CCache> deserialize(std::span<uint8_t> serializedCache);
+
+		private:
+			// we only do lookups based on main file contents + compiler options
+			struct Hash
+			{
+				inline size_t operator()(const SEntry& entry) const noexcept
+				{
+					std::vector<char> hashable = entry.compilerData.getHashable();
+					std::string hashableString(hashable.begin(), hashable.end());
+					hashableString += entry.mainFileContents;
+					return std::hash<std::string>{}(hashableString);
```

Comment:
no, you're supposed to precompute the hash and store it in `SEntry` for later retrieval here

because otherwise everytime you lookup, you end up hashing it